Add VoF Interface tracking

class4kayaker opened the pull request:
This is an implementation of the VoF based interface tracking implemented in #823, modified to link more closely with the main model. Currently the majority of the logic is implemented in the Simulator class due to relative ease of initial implementation. Moving most of these into a separate encapsulating class is planned, though some modification of approach may be necessary. Some of the modifications may be dependent on whether and how the results of the implementation are desired by the main model.

Current features:
- Signed distance Level Set or composition function initialization
- Dimensionally split, explicit update based on computed velocity field (Standard for VoF based methods due to problems from double fluxing induced by geometric flux calculation)
- Creation of internal per cell LS field for the interface reconstruction, primarily useful for visualization of reconstructed interface (left over from considering storing the reconstructed interface as a LS, not currently used for computation)

Current limitations:
- Only functional for 2D
- Can only handle parallelogram meshes
- Cannot handle variations in refinement level

Planned improvements:
- Make interface LS data available to MaterialModel (as composition?)

Possible improvements:
- Couple to level set method
- Reduce computation to narrow band around interface (Probably not worth additional code complexity)

bangerth commented:
@class4kayaker -- I think all of us who could review the patch are unsure about what exactly the VoF method is, and that makes it difficult to review the code. Could you prepare maybe 8-10 slides for when we meet, for example on Sunday in which you explain what the motivation for this is, and what the method is and does (in mathematical terms). Feel free to add a few pictures if you want. I think if we understand what VoF is, we'll have an easier time understanding what the implementation does.

-- thanks!

class4kayaker commented:
Sure, I will put together a presentation.

To have a basic explanation here, the volume of fluid method tracks the fluid interface by using an indicator function. At this point, this is similar to the current composition variable, but the volume of fluid method instead of using continuity as an approximation uses a piecewise constant approximation with a simple boundary. The basic volume of fluid method obtains the necessary information to generate the linear interface approximation by examining the surrounding cells. For the particular implementation I created, the interface is assumed to be linear on the unit cell, and is derived without an iterative method to second order accuracy, assuming correct volume fractions, by using the row an column sums on the 3x3 surrounding area. The advection is then handled by using method of characteristics to identify the region being advected through the interface.

The resulting algorithm, unlike a level set method, does not require frequent expensive reinitialization and can ensure volume conservation.

bangerth commented on include/aspect/parameters.h:
```diff
@@ -306,6 +306,19 @@ namespace aspect
+     * @{
+     */
+    bool                           vof_tracking_enabled;
+    double                         voleps;
```
can you choose a more descriptive name for this parameter?

bangerth commented on include/aspect/postprocess/vof_statistics.h:
```diff
@@ -0,0 +1,81 @@
+  {
+
+    /**
+     * A postprocessor that computes some statistics about the velocity.
```
probably outdated comment

bangerth commented on include/aspect/postprocess/visualization/vof_values.h:
```diff
@@ -0,0 +1,82 @@
+    namespace VisualizationPostprocessors
+    {
+      template <int dim>
+      class VoFValues
```
documentation missing

bangerth commented on include/aspect/simulator.h:
```diff
@@ -469,6 +481,92 @@ namespace aspect
 
+      // TODO: Identify whether and how the following functions can be moved
+      // out of the Simulator class, or which should be integrated as core
+      // functionality
```
Yes. I think it would be really nice if all of the VoF functionality could be separated into a class of its own, similar to the `FreeSurface` class, for example. We don't want to put every possible functionality into the `Simulator` class, as it would quickly become rather unreadable.

bangerth commented on include/aspect/simulator_signals.h:
```diff
@@ -80,7 +80,8 @@ namespace aspect
      */
-    boost::signals2::signal<void (std::vector<VariableDeclaration<dim> > &)>
+    boost::signals2::signal<void (const Parameters<dim> &parameters,
+                                  std::vector<VariableDeclaration<dim> > &)>
```
That's the wrong way to go. If a function you want to call requires access to the parameters object, just pass a bound pointer to a function, e.g.
```
  edit_finite_element_variables.connect (std_cxx11::bind (&MyClass::my_function,
                                                                                            std_cxx11::cref(parameters),
                                                                                            std_cxx11::_1));
```
This requires that you have access to the parameters object where you call the `connect` function, but this way you won't have to adjust the argument list of signals all the time.

bangerth commented on include/aspect/vof_initial_conditions/interface.h:
```diff
@@ -0,0 +1,211 @@
+         * Initialization data is a composition value between 0 and 1 at all
+         * points and should be accumulated by integration.
+         */
+        Compos,
```
Spell it out: `Composition`

bangerth commented on include/aspect/vof_initial_conditions/interface.h:
```diff
@@ -0,0 +1,211 @@
+         * Initialization data is an interface defined by a signed distance
+         * level set with positive value indicating fluid presence.
+         */
+        SDist_LS
```
Same here: `SignedDistanceLeastSquares` or `signed_distance_least_squares` (we tend to name the values of enums in lower case and underscore)

bangerth commented on include/aspect/vof_initial_conditions/interface.h:
```diff
@@ -0,0 +1,211 @@
+         * data.
+         */
+        virtual
+        unsigned int n_samp () const = 0;
```
spell it out: `n_samples`

bangerth commented on include/aspect/vofinterface/vof_refinement.h:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * A class that implements a mesh refinement criterion that refines the
+     * mesh in the uppermost nodes. This is useful for cases where one wants
+     * to accurately model processes at of close to the surface of the model.
```
update comment

bangerth commented on include/aspect/vofinterface/vof_system.h:
```diff
@@ -0,0 +1,148 @@
+/*
+ Copyright (C) 2016 by the authors of the ASPECT code.
```
I would prefer if this went into a directory `include/aspect/vof`

bangerth replied:
also, if the directory already contains the name `vof`, no need to duplicate this in the name of the file.

bangerth commented on include/aspect/vofinterface/vof_system.h:
```diff
@@ -0,0 +1,148 @@
+
+  namespace internal
+  {
+    namespace Assembly
```
I suspect that you only need these things in the VoF method. Maybe move all of it into a namespace `internal::VoF`?

bangerth commented on include/aspect/vofinterface/vof_system.h:
```diff
@@ -0,0 +1,148 @@
+           * each possible face or subface of the cell.
+           **/
+          std::vector<Vector<double>>     local_f_rhs;
+          std::vector<FullMatrix<double>> local_f_matrices_ext_ext;
```
maybe rename to `local_face_...` to make it clearer to the reader what the `f` stands for. If you use an IDE, you can do this renaming via refactoring rather than search/replace

bangerth replied:
also, we don't require C++11 yet, so you need to have a space between the two closing `>>` of the template argument list

bangerth commented on include/aspect/vofinterface/vof_system.h:
```diff
@@ -0,0 +1,148 @@
+           * to non-existent subfaces; or faces being assembled by the
+           * neighboring cell) are set to false.
+           **/
+          std::vector<bool>               assembled_rhs;
```
this variable is awkwardly named. I read this as "the values of the assembled rhs" whereas you mean it as a predicate ("which rhs contributions have been assembled?")

Maybe rename it to `face_rhs_contributions_mask`.

From a data efficiency perspective, you know the number of faces, so it may be more efficient to just store a fixed size array:
```
  bool assembled_rhs[GeometryInfo<dim>::faces_per_cell];
```
or
```
  std_cxx11::array<bool,GeometryInfo<dim>::faces_per_cell> assembled_rhs;
```

bangerth commented on include/aspect/vofinterface/vof_system.h:
```diff
@@ -0,0 +1,148 @@
+           * std::vector has length GeometryInfo<dim>::max_children_per_face *
+           * GeometryInfo<dim>::faces_per_cell.
+           **/
+          std::vector<std::vector<types::global_dof_index> >   neighbor_dof_indices;
```
same here: if you know the size of the array statically, make it a statically sized array

bangerth commented on include/aspect/vofinterface/vof_utils.h:
```diff
@@ -0,0 +1,63 @@
+
+namespace aspect
+{
+  namespace InterfaceTracker
```
I think it would be nice to name the namespace in correspondence to the method: `VoF` or `VolumeOfFluid`.

bangerth commented:
Overall points from a brief look at the implementation:
* It would be really nice if you could move all of this out of the `Simulator` class. I think the `FreeSurface` class is a good example of how almost everything can be externalized to a separate class and not very much has to be left in the `Simulator` itself. The author of `FreeSurface` is @ian-r-rose who I'm sure can answer questions about how the two classes interact.
* If I read this right, the implementation currently only handles a single field that is being tracked by the VoF method. I think it would be nice if that could be generalized to any number of fields, so that one can track, for example, where deep mantle and crust materials go over time.
* You use `[-0.5,-.5]^dim` as the unit cell. That is unusual: in deal.II, we always use `[0,1]^dim`, as defined in the `GeometryInfo` class. How much would it take to change this?

bangerth commented on include/aspect/vofinterface/vof_utils.h:
```diff
@@ -0,0 +1,63 @@
+
+    /**
+     * Function to calculate volume fraction contained by indicator function
+     * H(d-normal*x') on the [-0.5, 0.5]^dim unit cell
```
see overall comment on this pull request regarding the unit cell

gassmoeller commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,337 @@
+                         Patterns::Integer(0),
+                         "The number of fields to be handled usingg VoF interface tracking.");
+
+      prm.declare_entry ("Small volume", "1e-6",
```
This parameter name does not explain what it does. We have a similar parameter in the melt handler that is called "Melt scaling factor threshold", so what about "Volume threshold"?

gassmoeller commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,337 @@
+                         "the VoF system gets solved. See 'linear solver "
+                         "tolerance' for more details.");
+
+      prm.declare_entry ("VoF field names", "",
```
Can we use this and the next parameter from the "Compositional fields" subsection? That is how we do it for active particles.

gassmoeller commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,337 @@
+  void
+  VoFHandler<dim>::parse_parameters (ParameterHandler &prm)
+  {
+    prm.enter_subsection ("VoF config");
```
If you reuse existing parameters a lot of this function can be removed.

gassmoeller commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,337 @@
+                         Patterns::Double (0, 1),
+                         "Minimum significant volume. VOFs below this considered to be zero.");
+
+      prm.declare_entry ("VoF solver tolerance", "1e-12",
```
I guess it would be better to have separate solver tolerances for each compositional field, then you could just use the `Composition solver tolerance` for the VoF fields instead of introducing this new parameter. But until that is possible, this will be ok.

gassmoeller commented on source/vof/reconstruct.cc:
```diff
@@ -0,0 +1,389 @@
+
+    // Boundary reference
+    typename DoFHandler<dim>::active_cell_iterator endc =
+      sim.dof_handler.end ();
```
I know this class has direct access to the simulator, but I would still ask you to replace `sim.` by `this->get_...` wherever possible. There are three reasons: First, it is simpler for others to read, because most of the code base uses SimulatorAccess to access Simulator variables. Second, it is simpler to find out which calls actually change members of Simulator. Third, it is safer, because you can not accidentally change things.

class4kayaker commented on include/aspect/mesh_refinement/vof_interface.h:
```diff
@@ -0,0 +1,68 @@
+/*
```
The two separate implementations were an artifact of retaining a working earlier mesh refinement method while implementing a better approach. Since I have implemented a correct plugin based on flagging cells, I have removed the older method.

gassmoeller commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,358 @@
+    // Do checks on required assumptions
+    AssertThrow(dim==2,ExcMessage("Volume of Fluid Interface Tracking is currently only functional for dim=2."));
+    AssertThrow(this->get_parameters().CFL_number < 1.0, ExcMessage("Volume of Fluid Interface Tracking requires CFL < 1."));
+
```
Add a check that melt is not enabled.

gassmoeller replied:
Add an `AssertThrow(dynamic_cast<MappingCartesian<dim> *> (&this->get_mapping()) != 0, ExcMessage(...))`

gassmoeller commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,358 @@
+    if ( this->get_parameters().initial_adaptive_refinement > 0 ||
+         this->get_parameters().adaptive_refinement_interval > 0 )
+      {
+        // AMR active so check refinement strategy includes 'vof boundary'
```
needs to be implemented

gassmoeller commented on source/vof_initial_conditions/function.cc:
```diff
@@ -0,0 +1,132 @@
+                            "When set to true, initialization will be assumed to be a"
+                            "signed distance level set function.");
+
+          prm.declare_entry ("Number initialization samples", "3",
```
Move this into VoFHandler, and remove the above parameter. All existing InitialComposition plugins can be used for the initialization samples method, and all new plugins (derived from VoFInitialComposition) will implement the signed distance function.

gassmoeller commented on include/aspect/vof/utilities.h:
```diff
@@ -0,0 +1,117 @@
+                       const double vol);
+
+    /**
+     * Function to calculate xFEM-like function values on the unit cell.
```
What is xFEM? Please extend

naliboff commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,385 @@
+        const unsigned int vof_block_idx = data[f].volume_fraction.block_index;
+        const unsigned int vofN_block_idx = data[f].reconstruction.block_index;
+
+        // Reset current base to values at beginning of timestep
```
Remove comment.

naliboff commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,385 @@
+
+        // Reset current base to values at beginning of timestep
+
+        // Due to dimensionally split formulation, use strang splitting
```
change to `Strang (second-order dimensional) splitting`.

naliboff commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,385 @@
+        // Reset current base to values at beginning of timestep
+
+        // Due to dimensionally split formulation, use strang splitting
+        // TODO: Reformulate for unsplit (may require flux limiter)
```
Remove this comment and make an issue with a checklist of all the features to add to the VOF code after this pull request is merged.

naliboff commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,385 @@
+        // Due to dimensionally split formulation, use strang splitting
+        // TODO: Reformulate for unsplit (may require flux limiter)
+        bool update_from_old = true;
+        for (unsigned int dir = 0; dir < dim; ++dir)
```
For clarity rename `dir` to `direction`.

naliboff commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,385 @@
+        for (unsigned int dir = 0; dir < dim; ++dir)
+          {
+            // Update base to intermediate solution
+            if (!vof_dir_order_dsc)
```
Please rename `vof_dir_order_dsc` to  `direction_order_descending`.

naliboff replied:
Check whether this variable is stored during restart. Look in `particle_handler.h` at the function `serialize` for a template on how to do this.

naliboff commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,385 @@
+      {
+        const unsigned int c_var_index = this->introspection().compositional_index_for_name(iter->first);
+        const typename Simulator<dim>::AdvectionField adv_f = Simulator<dim>::AdvectionField::composition(c_var_index);
+        const VoFField<dim> vof_f= get_field(iter->second);
```
add a space between `vof_f` and `=`.

naliboff replied:
Also rename `vof_f` to `vof_field`.

naliboff commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,385 @@
+         iter!=vof_composition_map_index.end(); ++iter)
+      {
+        const unsigned int c_var_index = this->introspection().compositional_index_for_name(iter->first);
+        const typename Simulator<dim>::AdvectionField adv_f = Simulator<dim>::AdvectionField::composition(c_var_index);
```
Rename `adv_f` to `advection_field`.

naliboff commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,385 @@
+    for (std::map<std::string, unsigned int>::const_iterator iter=vof_composition_map_index.begin();
+         iter!=vof_composition_map_index.end(); ++iter)
+      {
+        const unsigned int c_var_index = this->introspection().compositional_index_for_name(iter->first);
```
Rename `c_var_index` to `composition_index`.

naliboff commented on source/vof/handler.cc:
```diff
@@ -0,0 +1,385 @@
+        const unsigned int c_var_index = this->introspection().compositional_index_for_name(iter->first);
+        const typename Simulator<dim>::AdvectionField adv_f = Simulator<dim>::AdvectionField::composition(c_var_index);
+        const VoFField<dim> vof_f= get_field(iter->second);
+        update_vof_composition(adv_f, vof_f, sim.solution);
```
Rename `adv_f` to `advection_field`.

class4kayaker commented:
@gassmoeller I believe I have covered your suggested modifications.

gassmoeller commented:
/run-tests

gassmoeller commented on include/aspect/mesh_refinement/volume_of_fluid_interface.h:
```diff
@@ -0,0 +1,75 @@
+
+
+#ifndef __aspect_mesh_refinement_volume_of_fluid_interface_h
+#define __aspect_mesh_refinement_volume_of_fluid_interface_h
```
Please use single underscores for all include guards in all header files (we removed the double underscores, as they are reserved for compiler purposes).

gassmoeller commented on include/aspect/postprocess/visualization/volume_of_fluid_values.h:
```diff
@@ -0,0 +1,121 @@
+
+
+#ifndef __aspect__postprocess_visualization_volume_of_fluid_values_h
+#define __aspect__postprocess_visualization_volume_of_fluid_values_h
```
same here

gassmoeller commented on include/aspect/postprocess/visualization/volume_of_fluid_values.h:
```diff
@@ -0,0 +1,121 @@
+    {
+      /**
+       * A class derived that implements a function that provides volume of
+       * fluid interface tracking data for graphical output.
```
Please make this comment more specific. What kind of data? It should make clear, why the class is called VolumeOfFluidValues.

gassmoeller commented on include/aspect/postprocess/volume_of_fluid_statistics.h:
```diff
@@ -0,0 +1,54 @@
+
+
+#ifndef __aspect__postprocess_volume_of_fluid_statistics_h
+#define __aspect__postprocess_volume_of_fluid_statistics_h
```
and here

gassmoeller commented on include/aspect/simulator_access.h:
```diff
@@ -690,6 +691,12 @@ namespace aspect
       get_melt_handler () const;
 
+      /**
+       * Return a pointer to the VolumeOfFluid handler.
```
Return a reference to ...

gassmoeller commented on include/aspect/volume_of_fluid/assembly.h:
```diff
@@ -0,0 +1,220 @@
+ */
+
+#ifndef __aspect__volume_of_fluid_assembly_h
+#define __aspect__volume_of_fluid_assembly_h
```
and here

gassmoeller commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,667 @@
+  namespace Assemblers
+  {
+    template <int dim>
+    void VolumeOfFluidAssembler<dim>::set_volume_fraction_threshold(const double value)
```
Where is this ever called? If it is necessary, would it be possible to get this from the vof_handler, instead of storing it in this assembler (i.e. `this->get_volume_of_fluid_handler().get_volume_fraction_threshold()`)?

gassmoeller commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,667 @@
+#include <aspect/volume_of_fluid/field.h>
+#include <aspect/volume_of_fluid/assembly.h>
+
+//#include <deal.II/base/quadrature_lib.h>
```
Remove these old includes

gassmoeller commented on source/simulator/solver.cc:
```diff
@@ -22,6 +22,7 @@
 #include <aspect/simulator.h>
 #include <aspect/global.h>
 #include <aspect/melt.h>
+#include <aspect/volume_of_fluid/handler.h>
```
you do not seem to use this include

gassmoeller replied:
Still here, please remove.

gassmoeller commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,667 @@
+          for (unsigned int k=0; k<volume_of_fluid_dofs_per_cell; ++k)
+            scratch.phi_field[k] = scratch.finite_element_values[solution_field].value(main_fe.component_to_system_index(solution_component, k), q);
+
+          // Init required local time
```
Please update comment

gassmoeller commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,667 @@
+
+      for (unsigned int face_no = 0; face_no < GeometryInfo<dim>::faces_per_cell; ++face_no)
+        {
+          const unsigned int f_dim = face_no/2; // Obtain dimension
```
Please move this comment to the line before, and explain this in more detail. Why is the dimension the face number divided by 2? Is this how deal.II orders faces?

gassmoeller commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,667 @@
+      const bool old_velocity_avail = (this->get_timestep_number() > 0);
+
+      const unsigned int f_dim = face_no/2; // Obtain dimension
+      const bool f_dir_pos = (face_no%2==1);
```
Please extend the comment, and find a better name for this variable. What is `f_dir_pos`?

gassmoeller commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,667 @@
+      const unsigned int n_f_q_points    = scratch.face_finite_element_values.n_quadrature_points;
+
+      // vol fraction and interface values are constants, so can set from first value
+      const double cell_vol = cell->measure();
```
As far as I remember `cell->measure()` is only an approximation of the cell volume. It might be exact for linear mappings though, so maybe it is find as is. Anyway, where do you need it, and could you easily replace it by a sum over `JxW(q)`, which gives you a better approximation for the real volume?

gassmoeller commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,667 @@
+                                       scratch.neighbor_i_d_values);
+                }
+
+              const double neighbor_vol = neighbor->measure();
```
same comment applies here

gassmoeller commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,667 @@
+                }
+
+              // Due to inability to reference this cell's values at the interface,
+              // need to do explicit calculation
```
Lets talk about this in more detail. What can you not access, and what are you calculating here?

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,681 @@
+    using namespace dealii;
+
+    template<>
+    double compute_fluid_fraction<2> (Tensor<1, 2, double> normal,
```
make both arguments `const`

gassmoeller commented on include/aspect/volume_of_fluid/assembly.h:
```diff
@@ -0,0 +1,220 @@
+        void set_volume_fraction_threshold(const double value);
+
+      private:
+        double volume_fraction_threshold;
```
Couldnt you call `this->get_volume_of_fluid_handler().get_volume_fraction_threshold` instead of copying the information here? That would be safer.

gassmoeller commented on include/aspect/volume_of_fluid/handler.h:
```diff
@@ -0,0 +1,225 @@
+
+
+    private:
+      // Parent simulator
```
/** ... */

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,149 @@
+  {
+    using namespace dealii;
+
+    // Utilitiess
```
typo

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,149 @@
+     */
+    template<int dim>
+    double compute_interface_location (const Tensor<1, dim, double> normal,
+                                       const double vol);
```
vol -> volume

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,149 @@
+     * @param values Values of the constructed polynomial at the specified points
+     */
+    template<int dim>
+    void xFEM_Heaviside_d_d(const int degree,
```
-> xFEM_Heaviside_derivative_d

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,149 @@
+
+
+    /**
+     * Function to do newton iteration calculation of correct d for a given
```
newton -> Newton

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,149 @@
+     *
+     * @param degree Maximum degree for exact integration
+     * @param normal Interface normal vector, pointed away from the included region
+     * @param d Interface parameter specifying location of interface in unit cell
```
Please specify that this is measured from the cell center

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,149 @@
+
+    /**
+     * Function to calculate volume fraction contained by indicator function
+     * H(d-normal*(x'-x_{cen}')) on the [0, 1]^dim unit cell where x_{cen} is
```
Please enclose all formulas with `$ ... $`

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,149 @@
+    template<int dim>
+    double calc_volume_of_fluid_flux_edge (const unsigned int compute_direction,
+                                           const double time_direction_derivative,
+                                           const Tensor<1, dim, double> inteface_normal_in_cell,
```
inteface -> interface

gassmoeller commented on source/volume_of_fluid/handler.cc:
```diff
@@ -0,0 +1,364 @@
+                        "set`` is specified, the intial data will be assumed to "
+                        "be in the form of a function which is positive when in "
+                        "the fluid, negative outside, and zero on the interface "
+                        "with the gradient being of the same magnitude "
```
gradient one everywhere

class4kayaker commented on tests/vof_circle_1/screen-output:
```diff
@@ -0,0 +1,398 @@
+
```
Reduce number of iterations for tests

class4kayaker commented:
I would appreciate a continued code review indicating what might need to be changed or fixed for inclusion.

Thank you.

gassmoeller commented:
Hi Jon, I will try to get another review done by the end of the week.

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+  {
+    using namespace dealii;
+
+    // Utilities
```
Please remove this line or even better make this a `namespace Utilities` just as we now have `MaterialModel::Utilities`

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+     */
+    template<int dim>
+    double compute_fluid_fraction (const Tensor<1, dim, double> normal,
+                                   const double d);
```
d -> distance?

class4kayaker replied:
In effect yes, but in practice d is more accurate due being defined in terms of the $d$ value required for the line $normal dot x = d$ to match the interface.

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+     * @param values Values of the constructed polynomial at the specified points
+     */
+    template<int dim>
+    void xFEM_Heaviside(const int degree,
```
The name of this function does not really explain what it does. What about `Heaviside_equivalent_polynomial_values`? Does this capture the meaning better? If you have a better idea feel free to go for it.

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+     * @param normal Interface normal vector, pointed away from the included region
+     * @param d Interface parameter specifying location of interface in unit cell
+     * @param points Locations to evaluate the constructed polynomial
+     * @param values Values of the constructed polynomial at the specified points
```
This documentation is identical to the one above, please clarify what is different about the two functions.

class4kayaker replied:
Re-read the documentation. One is for the equivalent to the heaviside, the other is the derivative.

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+     * @param values Values of the constructed polynomial at the specified points
+     */
+    template<int dim>
+    void xFEM_Heaviside_derivative_d(const int degree,
```
same name confusion as above, is there a better name?

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+    template<int dim>
+    double compute_interface_location_newton(const int degree,
+                                             const Tensor<1, dim, double> normal,
+                                             const double vol_frac,
```
Please replace any remaining location of `vol_frac` with `volume_fraction`, the few saved characters are not worth the mental effort to replace this every time someone reads it.

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+     * @param weights JxW values to use for quadrature
+     */
+    template<int dim>
+    double compute_fluid_volume_xFEM(const int degree,
```
what is the meaning of `xFEM`? would `compute_fluid_volume` be sufficient as a name?

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+    template<int dim>
+    double calc_volume_of_fluid_flux_edge (const unsigned int compute_direction,
+                                           const double time_direction_derivative,
+                                           const Tensor<1, dim, double> interface_normal_in_cell,
```
you can remove the `double` template parameter

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+     * the timestep.
+     */
+    template<int dim>
+    double calc_volume_of_fluid_flux_edge (const unsigned int compute_direction,
```
what is the meaning of `edge`? would `calculate_volume_flux` already capture all necessary information?

gassmoeller commented on include/aspect/volume_of_fluid/utilities.h:
```diff
@@ -0,0 +1,152 @@
+    /**
+     * Function to calculate flux volume fraction based on a method of
+     * characteristics approximation of the interface on the cell's face over
+     * the timestep.
```
Please add documentation for the parameters here. I am reasonably sure I know what `compute_direction` does, but a new reader would not, and I am entirely clueless what `time_direction_derivative` means :smile:.

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -999,6 +1010,18 @@ namespace aspect
         coupling[x.compositional_fields[0]][x.compositional_fields[0]] = DoFTools::always;
+
+      // If we are using VolumeOfFluid interface tracking, create a matrix block in the
+      // field corresponding to the volume fraction.
```
This comment does not fully explain why you need this. Why can you not use the same matrix block that is usually used for compositional fields? Is that one already used for something else (some of the other parameters you use)? If not, we might not need this matrix block and could instead use the existing one. We implemented this a while ago, because having many compositional fields can become the main memory limitation for a model. Please update the comment.

class4kayaker replied:
I have now implemented an approach similar to the current one for the compositional fields, but using the VoF 0 matrix.

gassmoeller replied:
Very nice! Thanks!

gassmoeller commented on source/simulator/simulator_access.cc:
```diff
@@ -617,6 +617,15 @@ namespace aspect
+  const VolumeOfFluidHandler<dim> &
+  SimulatorAccess<dim>::get_volume_of_fluid_handler () const
+  {
+    Assert (simulator->volume_of_fluid_handler.get() != 0,
```
replace 0 by nullptr

gassmoeller replied:
replace 0 by nullptr

gassmoeller commented on source/postprocess/visualization/volume_of_fluid_values.cc:
```diff
@@ -0,0 +1,201 @@
+          {
+            prm.enter_subsection("Volume of Fluid");
+            {
+              prm.declare_entry("Include internal reconstruction contour", "false",
```
I find the term `include` is not quite clear. What about `Output internal reconstruction contour`?

gassmoeller commented on source/postprocess/visualization/volume_of_fluid_values.cc:
```diff
@@ -0,0 +1,201 @@
+                                Patterns::Bool (),
+                                "Include the internal level set data use to save reconstructed interfaces");
+
+              prm.declare_entry("Include normals", "false",
```
Same here, what about `Output interface normals`?

gassmoeller commented on source/postprocess/visualization/volume_of_fluid_values.cc:
```diff
@@ -0,0 +1,201 @@
+
+              prm.declare_entry("Include normals", "false",
+                                Patterns::Bool (),
+                                "Include internal normal data in output (DEBUG)");
```
Why is this DEBUG? I think this might be useful. Please remove the DEBUG marker.

class4kayaker replied:
The main issue is that currently this does not adjust the output to match the real cell, but remains on the unit cell.

class4kayaker replied:
To clarify: if we were working on a mesh where the unit cell directions did not match the actual cell (eg due to rotations of the mesh), the normal vector would be in terms of the unit cell x and y, not the "physical space"

gassmoeller replied:
Ok, then write that into the documentation, but remove the debug. At the moment this can not happen anyway, as you only allow the box geometry. When this is extended to spherical models one just needs to find this place. Also add a comment like the following above the parameter:
`// TODO: Fix this for curved geometries`

gassmoeller commented on source/postprocess/visualization/volume_of_fluid_values.cc:
```diff
@@ -0,0 +1,201 @@
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(VolumeOfFluidValues,
+                                                  "volume of fluid values",
+                                                  "A visualization output object that outputs the  volume_of_fluid data."
```
I think it is not clear what `volume_of_fluid data` means in this context. What about `A visualization output object that outputs the volume fraction and optionally a level set field and the interface normal vectors of volume of fluid fields.`

gassmoeller commented on source/postprocess/visualization/volume_of_fluid_values.cc:
```diff
@@ -0,0 +1,201 @@
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(VolumeOfFluidValues,
+                                                  "volume of fluid values",
+                                                  "A visualization output object that outputs the  volume_of_fluid data."
+                                                  "Names are given in Postprocess/Visualization/Volume of fluid")
```
What do you mean by `names` here?

class4kayaker replied:
This is left over from when this was where the names of the fields were defined.

gassmoeller commented on source/postprocess/volume_of_fluid_statistics.cc:
```diff
@@ -0,0 +1,132 @@
+#include <aspect/postprocess/volume_of_fluid_statistics.h>
+#include <aspect/simulator_access.h>
+#include <aspect/volume_of_fluid/handler.h>
+#include <aspect/global.h>
```
the `simulator_access.h` and `global.h` include are already implicitly included via the header file, please remove them from here.

gassmoeller commented on source/postprocess/volume_of_fluid_statistics.cc:
```diff
@@ -0,0 +1,132 @@
+                                                                volume_of_fluid_values);
+                for (unsigned int q = 0; q < n_q_points; ++q)
+                  {
+                    //Use Kahan sum for improved consistency
```
Do you have evidence that this is actually necessary, i.e. did you observe problems with summation accuracy? We use simple addition in all of our other statistics postprocessors and I would like to keep the code as simple as possible.

class4kayaker replied:
Removed as requested.

gassmoeller commented on source/postprocess/volume_of_fluid_statistics.cc:
```diff
@@ -0,0 +1,132 @@
+                    //Use Kahan sum for improved consistency
+                    double cell_term = volume_of_fluid_values[q] * fe_values.JxW(q);
+                    double c_nterm = cell_term - volume_of_fluid_vol_corr;
+                    double nsum = volume_of_fluid_vol_sum + c_nterm;
```
make all of these doubles `const`

gassmoeller commented on source/postprocess/volume_of_fluid_statistics.cc:
```diff
@@ -0,0 +1,132 @@
+    ASPECT_REGISTER_POSTPROCESSOR(VolumeOfFluidStatistics,
+                                  "volume of fluid statistics",
+                                  "A postprocessor that computes some statistics about the "
+                                  "volume_of_fluid field.")
```
field -> fields

gassmoeller commented on source/simulator/core.cc:
```diff
@@ -366,6 +370,13 @@ namespace aspect
+    // for a mesh refinement strategy
+    if (parameters.volume_of_fluid_tracking_enabled)
+      {
+        volume_of_fluid_handler->initialize (prm);
```
I was a bit confused by you calling `initialize` here without ever calling `initialize_simulator` and `parse_parameters`, but that was before I saw that you do this in the constructor of the VoFHandler. Would there be a problem moving the calls to `initialize_simulator` and `parse_parameters` here, and move the initialization of the `assembler` member of VoFHandler into the `initialize` function instead of the constructor? That would be more consistent with the other plugin systems.

class4kayaker replied:
One issue is that for the current construction, the parameters must be parsed before the introspection object is created, which is done at construction of the Simulator object.

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+
+    initial_solution.compress(VectorOperation::insert);
+
+    sim.compute_current_constraints();
```
could you add a comment here of the following sort:
```suggestion
    // Apply constraints and update solution blocks. This is duplicated from
    // Simulator<dim>::set_initial_temperature_and_compositional_fields()
    // in order to separate the volume of fluid algorithm as much as possible from
    // the rest of the code.
    sim.compute_current_constraints();
```

Also I think it would be more efficient to move all of the `compress` and following operations into the `set_initial_volume_fractions` function above. Currently you reapply the current constraints for *all* fields every time you initialize *one* field. Additionally you would eliminate the duplication of this code between the compositional field and level set initialization functions.

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+  }
+
+  template <int dim>
+  void VolumeOfFluidHandler<dim>::initialize_from_composition_field (const VolumeOfFluidField<dim> field, const unsigned int f_ind)
```
please wrap parameters on separate lines

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+  }
+
+  template <int dim>
+  void VolumeOfFluidHandler<dim>::initialize_from_level_set (const VolumeOfFluidField<dim> field, const unsigned int f_ind)
```
please wrap parameters on separate lines

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+                                quadrature,
+                                update_JxW_values | update_quadrature_points);
+
+    double h = 1.0/n_init_samples;
```
can this be `const`?

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+    double h = 1.0/n_init_samples;
+
+    std::vector<types::global_dof_index>
+    local_dof_indicies (this->get_fe().dofs_per_cell);
```
```suggestion
    local_dof_indices (this->get_fe().dofs_per_cell);
```

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+        cell->get_dof_indices (local_dof_indicies);
+
+        cell_diam = cell->diameter();
+        d_func = this->get_initial_composition_manager().initial_composition(cell->barycenter(), f_ind);
```
if you move the two variables down here you can make them `const`

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+          {
+            volume_of_fluid_val = 0.0;
+          }
+        else
```
isnt this if-structure equivalent to:
```
if (d_func <=-0.5*cell_diam)
  ...
else if (d_func >= 0.5*cell_diam)
  ...
else
  ...
```

If so, please change to improve readability

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+                    double d = 0.0;
+                    Tensor<1, dim, double> grad;
+                    Point<dim> xU = quadrature.point (i);
+                    for (unsigned int di = 0; di < dim; ++di)
```
Please add a comment about the algorithm that happens in this loop, I do not quite understand what is happening (it is somehow related to the gradient in different directions?).

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+                        xH[di] += 0.5*h;
+                        xL[di] -= 0.5*h;
+                        double dH = this->get_initial_composition_manager().initial_composition(cell->intermediate_point(xH), f_ind);
+                        double dL = this->get_initial_composition_manager().initial_composition(cell->intermediate_point(xL), f_ind);
```
make the two doubles `const`

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+                        grad[di] = (dL-dH);
+                        d += (0.5/dim)*(dH+dL);
+                      }
+                    double ptvolume_of_fluid = VolumeOfFluid::Utilities::compute_fluid_fraction<dim> (grad, d);
```
make `const` and see my comment about this name above

gassmoeller commented on source/volume_of_fluid/setup_initial_conditions.cc:
```diff
@@ -0,0 +1,223 @@
+        initial_solution (local_dof_indicies[volume_of_fluid_ind]) = volume_of_fluid_val;
+      }
+
+    initial_solution.compress(VectorOperation::insert);
```
see my comment above about this part of the function

gassmoeller commented on source/volume_of_fluid/solver.cc:
```diff
@@ -0,0 +1,118 @@
+  template <int dim>
+  void VolumeOfFluidHandler<dim>::solve_volume_of_fluid_system (const VolumeOfFluidField<dim> field)
+  {
+    unsigned int block_idx = field.volume_fraction.block_index;
```
make `const`

gassmoeller commented on source/volume_of_fluid/solver.cc:
```diff
@@ -0,0 +1,118 @@
+                      << " iterations." << std::endl;
+
+    // Do not add VolumeOfFluid solver iterations to statistics, duplicaiton due to
+    // splitting messes with file format
```
Please explain a bit further `messes`, does it lead to duplicated or empty lines in the statistics file?
Also `duplicaiton`-> duplication

class4kayaker replied:
If I remember correctly, it produced "duplicated" lines where many of the statistics were missing (and possibly was inconsistent about which line had the actual data).

gassmoeller replied:
Then what about:
```
    // Do not add VolumeOfFluid solver iterations to statistics, because the repeated
    // execution for different directions within one time step duplicates lines in the output file.
```

gassmoeller commented on source/volume_of_fluid/solver.cc:
```diff
@@ -0,0 +1,118 @@
+    // Do not add VolumeOfFluid solver iterations to statistics, duplicaiton due to
+    // splitting messes with file format
+    // statistics.add_value("Iterations for VolumeOfFluid solver",
+    //                      solver_control.last_step());
```
remove these commented lines

class4kayaker commented on source/volume_of_fluid/handler.cc:
```diff
@@ -0,0 +1,364 @@
+                    ExcMessage("Volume of Fluid Interface Tracking requires that the 'volume of fluid interface' strategy be used for AMR"));
+
+        AssertThrow(this->get_parameters().adaptive_refinement_interval <(1/this->get_parameters().CFL_number),
+                    ExcMessage("Volume of Fluid Interface Tracking requires that the AMR interval be less than 1/CFL_number"));
```
The reconstruction algorithm currently used does not handle cells at different levels, and there would be a severe loss of accuracy even if the complex and novel logic necessary for handling that case was inserted. Using other reconstruction strategies might open this up later, but the current case is a major loss in accuracy at best.

gassmoeller replied:
Ok, then please add that to the message. Maybe:
```
Currently, the Volume of Fluid Interface Tracking requires that the AMR interval be less than 
1.0/CFL_number to preserve the intended accuracy. Circumventing this limitation requires
additional care in the reconstruction algorithm.
```

class4kayaker commented on source/volume_of_fluid/handler.cc:
```diff
@@ -0,0 +1,364 @@
+
+      prm.declare_entry ("Number initialization samples", "3",
+                         Patterns::Integer (1),
+                         "Number of sampled points per dimension when initializing from VOF");
```
Attempted clarification

class4kayaker commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,700 @@
+
+          double face_flux = 0;
+          double face_ls_d = 0;
+          double face_ls_time_grad = 0;
```
level set

tjhei replied:
longer variables are probably the better idea.

class4kayaker commented on source/volume_of_fluid/assembler.cc:
```diff
@@ -0,0 +1,700 @@
+          ||
+          (face->at_boundary() && cell->periodic_neighbor_is_coarser(face_no))
+          ||
+          (face->at_boundary() && neighbor->level() == cell->level() && neighbor->active()))
```
Note the `==` the logic also checks if the periodic neighbor cell in coarser, in which case that cell does the assembly.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+    typename DoFHandler<dim>::active_cell_iterator endc =
+      this->get_dof_handler().end ();
+
+    // Interface Reconstruction vars
```
Please be more specific here with a better variable name and more explanatory comment, neither `n_local` nor `Interface Reconstruction vars` tell me what this variable is for, and I do not know.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+    const unsigned int n_local = 9;
+
+    Vector<double> local_volume_of_fluids (n_local);
+    std::vector<Point<dim>> resc_cell_centers (n_local);
```
what does `resc` stand for? please rename variable, e.g. `reconstruction_cell_centers`?

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+    std::vector<Point<dim>> resc_cell_centers (n_local);
+    std::vector<typename DoFHandler<dim>::active_cell_iterator> neighbor_cells(n_local);
+
+    const unsigned int n_sums = 3;
```
Same here, what type of sums? Remember this code will be read by other people a lot of times, every minute we spend now on making it more readable will save others hours of work.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+    const unsigned int n_sums = 3;
+    std::vector<double> strip_sums (dim * n_sums);
+
+    const unsigned int n_normals = 6+1;
```
Same request, but additionally please add a comment about where the 6+1 comes from.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+                           update_JxW_values);
+
+    // Normal holding vars
+    Point<dim> uReCen;
```
is this `reconstructed_center_position`?

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+
+        // Obtain data for this cell and neighbors
+        cell->get_dof_indices (local_dof_indicies);
+        cell_volume_of_fluid = solution(local_dof_indicies[volume_of_fluid_ind]);
```
if possible make this `const double`

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+          {
+            initial_solution(local_dof_indicies[volume_of_fluid_ind]) = cell_volume_of_fluid;
+
+            //Identify best normal
```
Please unify spaces at the beginning of comments, I prefer a space, but either is fine as long as it is uniform.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+
+            //Identify best normal
+            // Get references to neighboring cells
+            for (unsigned int i = 0; i < 3; ++i)
```
What is 3? `dim`?  More likely it should be `n_sums`? Please avoid magic numbers like this to avoid possible bugs in the future.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+                  }
+                else
+                  cen = cell;
+                for (unsigned int j = 0; j < 3; ++j)
```
Please put an empty line between the else and the for-loop otherwise this is confusing to read. Also replace 3 by an appropriate variable name. Finally, add a comment: What is `curr` below, in particular compared to `cen` above?

class4kayaker replied:
Comments added to attempt clarification

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+
+            // std::cout << "  Strip sums: " << std::endl;
+            // for (unsigned int i = 0; i < dim * n_sums; ++i)
+            //   std::cout << "    " << strip_sums[i] << std::endl;
```
remove the commented code

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+            //   std::cout << "    " << strip_sums[i] << std::endl;
+
+            // Calculate normal vectors for the 6 candidates from the efficient
+            // least squares approach
```
This comment needs to be extended significantly and the individual cases in the loop below documented. What are the 6 candidates and why are there 6? How does the efficient least squares approach work (in simple terms)? And what are `di` and `di2`?

class4kayaker replied:
Added attempted clarification

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+                normals[6][1] = 0;
+              }
+
+            unsigned int mn_ind = 0;
```
what does `mn_ind` stand for? please rename

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+
+                  if (normal_norm > volume_fraction_threshold) // If candidate normal too small set error to maximum
+                    {
+                      d_vals[nind] = VolumeOfFluid::Utilities::compute_interface_location_newton<dim> (max_degree, normals[nind], cell_volume_of_fluid, cell_vol,
```
wrap line by putting arguments on separate lines

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+                  mn_ind = nind;
+                // std::cout << "\t" << normals[nind] << " d ";
+                // std::cout << d_vals[nind] << " e ";
+                // std::cout << errs[nind] << " " << mn_ind << std::endl;
```
Please remove commented code.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+        //   {
+        //     normal[0] = 0.0;
+        //     normal[1] = 0.0;
+        //   }
```
same here

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+        for (unsigned int i=0; i<n_volume_of_fluidLS_dofs; ++i)
+          {
+            // Recenter unit cell on origin
+            Tensor<1, dim, double> uSupp = volume_of_fluidLS_var.fe->unit_support_point(i)-uReCen;
```
Rename uSupp to something for which I can understand what it is. `recentered_support_point`?

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+    sim.compute_current_constraints();
+    sim.current_constraints.distribute(initial_solution);
+
+    // solution.block(volume_of_fluid_blockidx) = initial_solution.block(volume_of_fluid_blockidx);
```
why is this not needed? if not needed please remove.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+  }
+
+  template <>
+  void VolumeOfFluidHandler<2>::update_volume_of_fluid_composition (const typename Simulator<2>::AdvectionField composition_field,
```
hand over `compositional_field` by reference to avoid copies

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+
+  template <>
+  void VolumeOfFluidHandler<2>::update_volume_of_fluid_composition (const typename Simulator<2>::AdvectionField composition_field,
+                                                                    const VolumeOfFluidField<2> volume_of_fluid_field,
```
same here

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+                                                                    const VolumeOfFluidField<2> volume_of_fluid_field,
+                                                                    LinearAlgebra::BlockVector &solution)
+  {
+    const int dim = 2;
```
```suggestion
    const unsigned int dim = 2;
```

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+    initial_solution.reinit(sim.system_rhs, false);
+
+    // Normal holding vars
+    Point<dim> uReCen;
```
the same comments about naming that I had in the function above apply here

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+
+    const FiniteElement<dim> &system_fe = this->get_fe();
+
+    std::vector<types::global_dof_index> local_dof_indicies (system_fe.dofs_per_cell);
```
```suggestion
    std::vector<types::global_dof_index> local_dof_indices (system_fe.dofs_per_cell);
```

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+        double normall1n = 0.0;
+        for (unsigned int i=0; i<dim; ++i)
+          {
+            normall1n += numbers::NumberTraits<double>::abs(normal[i]);
```
is `std::abs` not working here?

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+                                                  .component_to_system_index(volume_of_fluidN_c_index+i, 0)]);
+
+        Tensor<1, dim, double> nnormal;
+        double normall1n = 0.0;
```
```suggestion
        double normal_l1_norm = 0.0;
```
This took me a minute to figure out, and everyone else who will read this code as well.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+          }
+        if (normall1n > volume_fraction_threshold)
+          {
+            nnormal = normal / normall1n;
```
hmm, so `nnormal` is an L1-normalized normal vector? That is a somewhat unusual choice, please document why you do that.

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+            nnormal = normal / normall1n;
+          }
+        //Calculate correct factor to retain vol frac and [0,1] bound
+        double fact = 2.0*(0.5-abs(cell_volume_of_fluid-0.5));
```
```suggestion
        double correction_factor = 2.0*(0.5-std::abs(cell_volume_of_fluid-0.5));
```

gassmoeller commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+              = system_fe.component_to_system_index(composition_field.component_index(sim.introspection),
+                                                    /*dof index within component*/i);
+
+            Tensor<1, dim, double> uSupp = support_points[i]-uReCen;
```
same name change as above

class4kayaker commented on source/volume_of_fluid/reconstruct.cc:
```diff
@@ -0,0 +1,488 @@
+
+    LinearAlgebra::BlockVector initial_solution;
+
+    sim.computing_timer.enter_section("   Reconstruct VolumeOfFluid interfaces");
```
See above (explicit indentation is required).

gassmoeller replied:
see answer above

gassmoeller commented on source/volume_of_fluid/assembly.cc:
```diff
@@ -0,0 +1,303 @@
+                                                                   const unsigned int dir,
+                                                                   const bool update_from_old)
+  {
+    sim.computing_timer.enter_section ("   Assemble volume of fluid system");
```
I stand corrected. However, only some sections have this indentation and it was likely an historical accident that was copy-pasted into several places. Therefore I opened #2781 to fix the places in master, and would still ask you to change this line :smile:.

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+        // Values calculated using sympy
+        if (d<-0.5*norm1)
+          {
+            for (int i =0; i < basis_count; ++i)
```
`unsigned int`

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+          {
+            // Full cell
+            coeffs[0] = 1.0;
+            for (int i =1; i < basis_count; ++i)
```
`unsigned int`

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+        else if (norm1< 1e-7)
+          {
+            coeffs[0] = 0.5;
+            for (int i =1; i < basis_count; ++i)
```
`unsigned`

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+        else if (d<=-triangle_break)
+          {
+            //Triangle
+            const double d_n = d + 0.5*norm1;
```
In the formulas below, are the `L`'s necessary? if note remove. And replace the 2 and 3 by 2.0 and 3.0 to avoid confusion.

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+        AssertThrow(degree<=max_degree,
+                    ExcMessage("Cannot generate xFEM polynomials are only functional for degrees<2."));
+
+        // Values calculated using sympy
```
The algorithm below needs some explanation in this comment. Keep the reference to sympy, but also explain what you are doing.

class4kayaker replied:
Clarification added

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+      }
+
+      template<>
+      void xFEM_Heaviside_derivative_d<2>(const int degree,
```
`const unsigned int`

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+
+
+      template<int dim>
+      double compute_interface_location_newton(const int degree,
```
`const unsigned int`

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+                                               const std::vector<double> &weights)
+      {
+        double norm1=0.0;
+        for (int i=0; i<dim; ++i) norm1+=fabs(normal[i]);
```
please wrap line

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+            xFEM_Heaviside_derivative_d(degree, normal, d_guess, points, df_values);
+
+            f_guess=0.0;
+            df_guess=0.0;
```
if you only use these variables within the loop then declare them in the loop.

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+      }
+
+      template<int dim>
+      double compute_fluid_volume(const int degree,
```
`unsigned`

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+
+        xFEM_Heaviside(degree, normal, d, points, f_values);
+
+        double vol=0.0;
```
Is this `volume_fraction`, or `cell_volume`? based on the loop below I suppose it is `volume_fraction`, but you used `vol` as abbreviation for `cell_volume` in the function above. This is an example for why we ask so persistent for self-explanatory variable names.

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+        Tensor<1, dim> i_normal;
+        double i_d;
+
+        i_d = d_face+0.5*timeGrad;
```
```suggestion
        const double i_d = d_face + 0.5 * time_derivative;
```

gassmoeller commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+
+      template<int dim>
+      double calculate_volume_flux(const unsigned int dir,
+                                   const double timeGrad,
```
Our [style guide](https://www.dealii.org/developer/doxygen/deal.II/CodingConventions.html) reserves CamelCase for classes and variables. What about `time_derivative`?

class4kayaker commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+          }
+        if (norm1 == 0.0)
+          {
+            // Case should never occur, return reasonable output for edge circumstance
```
The correct result is undefined. The main issue is to select some reasonable return should it occur.

class4kayaker commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+            return 0.5;
+          }
+        dtest = d / norm1;
+        mpos = 1.0 - max/norm1;
```
`mpos` is equal to the absolute value of the smaller vector entry in the 1-norm normalized normal vector, and is chosen due to making a the following formulas significantly simpler. I have added comments indicating this.

class4kayaker commented on source/volume_of_fluid/utilities.cc:
```diff
@@ -0,0 +1,687 @@
+        const int max_degree = 1;
+
+        AssertThrow(degree<=max_degree,
+                    ExcMessage("Cannot generate xFEM polynomials are only functional for degrees<2."));
```
Rephrased for clarity.

class4kayaker commented on tests/vof_linear_c.prm:
```diff
@@ -0,0 +1,99 @@
+# Test for linear interface under constant velocity
+# Error should be at machine precision
```
This may have been a remnant of when I did not necessarily generate a composition field for a VoF field, and was intended to test that code path.

gassmoeller replied:
Then feel free to remove this test. If we do not remember what it is for, it will only be confusing in the future.

class4kayaker commented:
I have added the doc/modules/changes entry, and attempted to fix most of your issues. I have a few cases where I would like some advice on whether my attempts were successful or the best way of clarifying the issue.

class4kayaker commented on tests/vof_err_calc.h:
```diff
@@ -0,0 +1,352 @@
+#include <deal.II/fe/fe_values.h>
+
+
+namespace aspect
```
Attempted to cover the issues.

bangerth commented:
@class4kayaker At some point you will have to rebase this patch. The branch point for your code is currently about 18 months old. Can you try to do a rebase? I would not be surprised if some adjustments will be necessary to make things work with the current top of the main branch, though github at least tells me that there are no merge conflicts.

gassmoeller commented:
I think we had a rebase closer than 18 months (last summer?), but anyway, yes a rebase would be good. I also allowed testing, so you can see the reference tester results and update you test reference output.

class4kayaker commented:
If I do a rebase, I will likely need to squash most or all of the commits due to many of the older implementation pieces having been more entangled with pieces of the main simulator code that are no longer relevant (but were essential to correct functioning at the time). As this would be somewhat irreversible, I would like to confirm that that would be acceptable before making the adjustment.

bangerth commented:
Try and see whether a regular rebase goes through. If not, squash it all into one. Then we can always separate things into different commits again at a later point that are based on logically independent pieces, rather than demonstrating a long and tortuous history.

class4kayaker commented:
Pushed a squashed merge.

class4kayaker commented:
I believe I have addressed the issues, and I have made a squashed rebase onto master. The unresolved issues from the latest review by @gassmoeller are ones where the modification could likely use external review to check whether they are sufficient, or additional clarification is needed.

gassmoeller commented:
Ok, thanks. I think it is time to come to an end with this PR and leave follow-up changes to the future. I opened a PR against your version to include some last formal changes, and left a TODO for you in the changelog entry. Please fill in a reference to your submitted paper, and let me know once you are done with that. Then we can merge this mega-project.

class4kayaker commented:
I have merged the PR you made. I am not sure what the procedure is for making references to submitted, but not yet final papers, so I would appreciate any advice there.

gassmoeller commented:
Thinking about it the changelog might not be the right place to put this anyway because it will become part of the release notes (and the preliminary reference to the paper will change). Please just remove the last half-sentence (the part about the reference). Do not forget to let us know once the paper is accepted so that we include the paper in the list of publications about ASPECT.

class4kayaker commented:
I have removed the paper reference.

gassmoeller commented:
Ok, thanks. From my side this is ready to merge, but since it is quite a significant addition, I asked @tjhei to have a quick look at the overall structure (files, and code that influences `Simulator`). Also as a follow-up (separate) PR it would be great to get a little more examples, e.g. a benchmark setup or a cookbook. I understand you have submitted a paper about it and it is part of your thesis, would you be willing to provide some examples in the future? Also, what are your plans for using/supporting this technique in the future? Have you plans for extending this (e.g. to 3D or curved geometries)? I am mostly asking to get a mental picture of a future roadmap for this part of the code.

class4kayaker commented:
I would be fine with writing up a cookbook.

I would be interested in working on extending/improving the implementation, but in the near future I have the work to finish up my PhD that I need to handle, so I do not have any plans for such on the near time horizon.

class4kayaker commented on source/simulator/parameters.cc:
```diff
@@ -1705,6 +1731,18 @@ namespace aspect
 
+    prm.enter_subsection ("Volume of Fluid");
+    {
+      const bool explicit_volume_of_fluid_tracking_enabled = prm.get_bool("Enable interface tracking");
```
I believe that this may have been left over from when this option indicated that it was necessary to load the VoF handler which then did all the parsing. I have checked, and since it does appear to be redundant, I have removed the parameter.

class4kayaker commented on tests/vof_err_calc.h:
```diff
@@ -0,0 +1,343 @@
+    using namespace dealii;
+
+    template <int dim>
+    class VoFMMSErr : public Interface<dim>, public ::aspect::SimulatorAccess<dim>
```
Renamed for clarification

tjhei commented:
great, thanks! I am going to merge, then.
